src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                    if (!TryGetOptions(firstBlock.Syntax.SyntaxTree, firstBlock.Language,
+                                       context.Options, supportsDiscard,
+                                       unusedValueExpressionStatementOption, unusedValueAssignmentOption,
+                                       context.CancellationToken, out var options))
+                    {
+                        return;
+                    }
+
+                    var blockAnalyzer = new BlockAnalyzer(symbolStartAnalyzer, options);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeExpressionStatement, OperationKind.ExpressionStatement);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeDelegateCreationOrAnonymousFunction, OperationKind.DelegateCreation, OperationKind.AnonymousFunction);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeConversion, OperationKind.Conversion);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeFieldOrPropertyReference, OperationKind.FieldReference, OperationKind.PropertyReference);
+                    context.RegisterOperationBlockEndAction(blockAnalyzer.AnalyzeOperationBlockEnd);
+
+                    return;
+
+                    // Local Functions.
+                    bool HasSyntaxErrors()
+                    {
+                        foreach (var operationBlock in context.OperationBlocks)
+                        {
+                            if (operationBlock.SemanticModel.GetSyntaxDiagnostics(operationBlock.Syntax.Span, context.CancellationToken).HasAnyErrors())
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+                }
+
+                private void AnalyzeExpressionStatement(OperationAnalysisContext context)
+                {
+                    if (_options.UnusedValueExpressionStatementPreference == UnusedValuePreference.None)
+                    {
+                        return;
+                    }
+
+                    Debug.Assert(_options.UnusedValueExpressionStatementSeverity != ReportDiagnostic.Suppress);
+
+                    // Bail out cases for report unused expression value:
+                    //  1. Null type
+                    //  2. Void returning method invocations
+                    //  3. Bool returning method invocations
+                    //  4. Constant expressions
+                    //  5. Assignments, increment/decrement operations
+                    //  6. Invalid operation
+                    //  7. Expression and expression statement have differing first tokens, indicating explicit discard (for example, VB call statement)
+                    //  8. Special cases: Methods beloning to System.Threading.Interlocked and System.Collections.Immutable.ImmutableInterlocked.
```

Comment:
this is an entire 'what list', but it doesn't explain why.  For example, why do we bail on bool-returning method invocations?  if i do `a.Equals(b);` isnt' that a sign of an issue?

Reply:
i'd love an explanation *in code* for why we're doing each of those cases.  If that means breaking it up into a comment/check/comment/check, that's even better as these massive expressions are such a pain to maintain :)